### PR TITLE
add replaceComponentRenderer browser api runner

### PR DESCRIPTION
**Pull request by stevensurgnier**
Happy Friday!

This PR adds a new browser api runner, `replaceComponentRenderer `. We're using it to choreograph transitions between "pages." The original idea came from #1415. Unlike #1415, I recommend we provide a more general solution and leave the complexity of caching the history of page props (i.e. previous and next) to the client.

here are the important bits
https://github.com/gatsbyjs/gatsby/pull/2557/files#diff-32927b6e4c908ebb2e2648dbb183c657
https://github.com/gatsbyjs/gatsby/pull/2557/files#diff-aa6c0e8762b0f602655114263902a258

### CODE REVIEW

File: packages/gatsby/cache-dir/component-renderer.js [hunk 1]
```diff
@@ -91,10 +92,12 @@ class ComponentRenderer extends React.Component {
   }
 
   render() {
+    const pluginResponses = apiRunner(`replacePageComponentRenderer`, { props: this.props })
+    const replacementPage = pluginResponses[0]
     // If page.
     if (this.props.page) {
```
Comment (KyleAMathews) [inline packages/gatsby/cache-dir/component-renderer.js]: Any reason not to just put `replacePageComponentRenderer` inside the page check?

Comment (stevensurgnier) [reply packages/gatsby/cache-dir/component-renderer.js]: That makes sense for our particular use case. We don't have any plans to transition layouts. However, I didn't want to make that assumptions for others. This seemed more general at the cost of clients (such as myself) having to check if it was a layout in the runner implementation.

**Comment by stevensurgnier** [comment]
I'll add an example site asap. My afternoon is booked though.

File: packages/gatsby/cache-dir/component-renderer.js (hunk 1 above)
Comment (stevensurgnier) [reply packages/gatsby/cache-dir/component-renderer.js]: Now that I've said that, it shouldn't really be named  replace**Page** then, huh?

Comment (stevensurgnier) [reply packages/gatsby/cache-dir/component-renderer.js]: You cool with both `replacePageComponentRenderer` and `replaceLayoutComponentRenderer`?

Comment (KyleAMathews) [reply packages/gatsby/cache-dir/component-renderer.js]: Haha yeah that's what tripped me up — `replaceComponentRenderer` would do nicely

**Comment by KyleAMathews** [comment]
Oh this needs documented also @ https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/utils/api-browser-docs.js

**Comment by stevensurgnier** [comment]
- [x] example site
- [x] docs

**Comment by KyleAMathews** [comment]
Perfect! Sorry for the slow response — been at conference past few days.

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 26d5f98a1be1ef39244714966d42e3ad31956529

https://app.netlify.com/sites/using-glamor/deploys/59f0f8c0df995314eba3b9ab

**Comment by tannerlinsley** [comment]
Hey, this stuff looks great! Glad you guys were able to figure out a nice API :) Great job.

**Comment by monsieurnebo** [comment]
Hi ! Thanks for this feature, it's working fine for pages, but it doesn't seem to fit with layouts usage.

Let's say we have a **layout** for header & footer, but the in between is the **page**. This transition only applies on the page, keeping the header & footer without transition. 

Am I forgetting something, or it's missing in this example?

**Comment by KyleAMathews** [comment]
@monsieurnebo layouts & pages are different. There's lots of use cases where you'd want to transition just the page and not the layout e.g. sliding between pages.

That being said, it'd be cool to support somehow optionally applying the transition to the entire page.

**Comment by monsieurnebo** [comment]
Yes, it definitly would be cool !

**Comment by b0gd4n** [comment]
is there an example using this?

**Pull request merged**